### PR TITLE
OM-1381 tooltip fixes

### PR DESCRIPTION
**Pull request by emessiha**
- Switching Education content fetch to JSON
- Updating Code tool tips and annotators to use Reshift Education content
- Implemented Global Caching service for Education content (plugin wide)

### CODE REVIEW

File: src/main/java/com/reshiftsecurity/education/DevContent.java [hunk 1]
```diff
@@ -20,20 +20,27 @@
...
+    public String getContent() { return content; }
+
+    public String getTitle() {
+        return title;
     }
 
-    public String getSectionTitle() {
-        return sectionTitle;
+    public String getContentWithTitle() {
+        return String.format("<h1>%s</h1>", this.title) +
```
Comment (qcorporation) [inline src/main/java/com/reshiftsecurity/education/DevContent.java]: escape title?

File: src/main/java/com/reshiftsecurity/education/EducationService.java [hunk 2]
```diff
@@ -20,59 +20,55 @@
...
+        } catch (IOException e) {
+            e.printStackTrace();
         }
+        Reader jsonReader = null;
+        try {
+            jsonReader = new InputStreamReader(input, "UTF-8");
+        } catch (UnsupportedEncodingException e) {
+            e.printStackTrace();
+        }
+        devContent = gson.fromJson(jsonReader, devContentType);
```
Comment (qcorporation) [inline src/main/java/com/reshiftsecurity/education/EducationService.java]: if you fall into the catch, do you really want to call gson.fromJson(...)?

File: src/main/java/com/reshiftsecurity/education/EducationService.java [hunk 3]
```diff
@@ -20,59 +20,55 @@
...
-            html = Jsoup.connect(bugEduUrl).get().html();
-            if (!StringUtils.isEmpty(html)) {
-                Whitelist htmlTagWhiteList = Whitelist.relaxed();
-                html = Jsoup.clean(html, htmlTagWhiteList);
-            }
-        } catch (IOException e) {
-            e.printStackTrace();
-        }
-        return html;
+        return eduBaseUri.resolve(String.format("/%s/data.json", vulnerabilityType)).toString();
```
Comment (qcorporation) [inline src/main/java/com/reshiftsecurity/education/EducationService.java]: same here, if catch the URISyntaxException do you really want to use eduBaseUri?

File: src/main/java/com/reshiftsecurity/plugins/intellij/gui/editor/BugAnnotator.java [hunk 4]
```diff
@@ -202,18 +202,20 @@ private static void addAnnotation(final ExtendedProblemDescriptor problemDescrip
...
-			buffer.append(ResourcesLoader.getString("findbugs.name")).append(": ").append(StringUtilFb.html2text(BugInstanceUtil.getBugPatternShortDescription(problemDescriptor.getBug().getInstance()))).append('\n');
-			buffer.append(StringUtilFb.html2text(BugInstanceUtil.getDetailText(problemDescriptor.getBug().getInstance())));
-			if (i < problemDescriptors.size() - 1) {
-				//noinspection HardcodedLineSeparator
-				buffer.append("\n\n");
-			}
-		}
-
+		buffer.append(_eduCacheService.getBriefOverview(problemDescriptors, true));
+//		for (//noinspection LocalCanBeFinal
```
Comment (qcorporation) [inline src/main/java/com/reshiftsecurity/plugins/intellij/gui/editor/BugAnnotator.java]: can we put a FIXME or TODO as to why this code is not removed and remains as comments

File: src/main/java/com/reshiftsecurity/plugins/intellij/gui/editor/BugsLineMarkerProvider.java [hunk 5]
```diff
@@ -191,24 +192,25 @@ public String fun(final PsiElement psiElement) {
...
-				if (i < problemDescriptors.size() - 1) {
-					buffer.append("<HR>");
-				}
-
-			}
-
-			buffer.append("</BODY></HTML>");
+			buffer.append(_eduCacheService.getBriefOverview(problemDescriptors, false));
+
+//			final int problemDescriptorsSize = problemDescriptors.size();
```
Comment (qcorporation) [inline src/main/java/com/reshiftsecurity/plugins/intellij/gui/editor/BugsLineMarkerProvider.java]: same as above

File: src/main/java/com/reshiftsecurity/plugins/intellij/gui/editor/BugAnnotator.java [hunk 6]
```diff
@@ -202,18 +202,20 @@ private static void addAnnotation(final ExtendedProblemDescriptor problemDescrip
...
-			final ExtendedProblemDescriptor problemDescriptor = problemDescriptors.get(i);
-			buffer.append(ResourcesLoader.getString("findbugs.name")).append(": ").append(StringUtilFb.html2text(BugInstanceUtil.getBugPatternShortDescription(problemDescriptor.getBug().getInstance()))).append('\n');
-			buffer.append(StringUtilFb.html2text(BugInstanceUtil.getDetailText(problemDescriptor.getBug().getInstance())));
-			if (i < problemDescriptors.size() - 1) {
-				//noinspection HardcodedLineSeparator
-				buffer.append("\n\n");
-			}
-		}
-
+		buffer.append(_eduCacheService.getBriefOverview(problemDescriptors, true));
```
Comment (qcorporation) [inline src/main/java/com/reshiftsecurity/plugins/intellij/gui/editor/BugAnnotator.java:207]: is this protected against xss?

File: src/main/java/com/reshiftsecurity/plugins/intellij/gui/editor/BugsLineMarkerProvider.java [hunk 7]
```diff
@@ -191,24 +192,25 @@ public String fun(final PsiElement psiElement) {
...
-				buffer.append("</H3>");
-				buffer.append(PATTERN.matcher(BugInstanceUtil.getDetailText(problemDescriptor.getBug().getInstance())).replaceAll(""));
-				if (i < problemDescriptors.size() - 1) {
-					buffer.append("<HR>");
-				}
-
-			}
-
-			buffer.append("</BODY></HTML>");
+			buffer.append(_eduCacheService.getBriefOverview(problemDescriptors, false));
```
Comment (qcorporation) [inline src/main/java/com/reshiftsecurity/plugins/intellij/gui/editor/BugsLineMarkerProvider.java:197]: is this protected against xss?

File: src/main/java/com/reshiftsecurity/plugins/intellij/service/EducationCachingService.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+public class EducationCachingService {
+    private List<VulnerabilityDetails> contentCache;
+    private final String SECTION_OVERVIEW="Overview";
+
+    public EducationCachingService() {
+        this.contentCache = new ArrayList<>();
+    }
+
+    private Optional<VulnerabilityDetails> getContentFromCache(String issueType) {
+        for (VulnerabilityDetails vulnerabilityDetails : contentCache) {
```
Comment (qcorporation) [inline src/main/java/com/reshiftsecurity/plugins/intellij/service/EducationCachingService.java:45]: just a nitpick (can ignore): consider using the streaming api
contentCache.stream().filter(...).findAny().orElse(Optional.empty());

**Review by qcorporation — APPROVED** [review]
besides those issues, looks ok

File: src/main/java/com/reshiftsecurity/education/EducationService.java (hunk 2 above)
Comment (emessiha) [reply src/main/java/com/reshiftsecurity/education/EducationService.java]: Good catch! no pun intended lol

File: src/main/java/com/reshiftsecurity/plugins/intellij/gui/editor/BugAnnotator.java (hunk 6 above)
Comment (emessiha) [reply src/main/java/com/reshiftsecurity/plugins/intellij/gui/editor/BugAnnotator.java:207]: Will add sanitization in `getBriefOverview`

File: src/main/java/com/reshiftsecurity/plugins/intellij/gui/editor/BugsLineMarkerProvider.java (hunk 7 above)
Comment (emessiha) [reply src/main/java/com/reshiftsecurity/plugins/intellij/gui/editor/BugsLineMarkerProvider.java:197]: should be addressed once sanitization is added to `getBriefOverview`

**Pull request merged**